File: python/11-20/Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @return a ListNode
    def removeNthFromEnd(self, head, n):
        su = ListNode(0)
        su.next=head
        pre,cur,p = su,head,head
        for i in range(0,n):
            p=p.next
        while p!=None:
            p=p.next
            cur=cur.next
            pre=pre.next
        pre.next=cur.next
        return su.next

File: Cpp/11-20/Remove Nth Node From End of List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 
 /*
    runtime: 4ms
    error:  0
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode thead(0);
        thead.next = head;
        ListNode* first = &thead,*second = &thead;
        int i=0;
        while(i<n){
            i++; 
            first = first->next;
        }
        while(first->next!=nullptr){
            first=first->next; 
            second = second->next;
        }
        first = second->next;
        second->next = second->next->next;
        delete first;
        return thead.next;
    }
};
